src/emailjs-tcp-socket.js

```diff
@@ -619,8 +662,16 @@
                 // process the data available on the socket
                 self._onData(readInfo.data);
 
-                // queue the next read
-                self._readLegacySocket();
+                // Queue the next read.
+                // If a STARTTLS handshake might be upcoming, postpone this onto
+                // the task queue so the IMAP client has a chance to call upgradeToSecure;
+                // without this, we might eat the beginning of the handshake.
+                // If we are already secure, just call it (for performance).
+                if (self.ssl) {     // are we secure yet?
+                    self._readLegacySocket();
+                } else {
+                    setZeroTimeout(self._readLegacySocket.bind(self));
```

Comment:
I think the main reason it works is this bit of code: https://github.com/emailjs/emailjs-tcp-socket/blob/master/src/emailjs-tcp-socket.js#L606-L609

Promise resolutions go on the microtask queue, whereas `setTimeout` and `postMessage` go on the macrotask queue. [(Good article on this here.)](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) This means that promise resolutions will always execute before the scheduled read. So the higher-up code will always have a chance to call `upgradeToSecure()` as long as it does it directly in the `ondata` callback, or a promise resolved from there.